store/postgres/src/entity_changes.rs

```diff
@@ -0,0 +1,156 @@
+            terminate_worker,
+            worker_barrier,
+            started: false,
+        }
+    }
+
+    /// Begins processing notifications coming in from Postgres.
+    pub fn start(&mut self) {
+        if !self.started {
+            self.worker_barrier.wait();
+            self.started = true;
+        }
+    }
+
+    fn listen(
+        url: String,
+    ) -> (
+        Receiver<EntityChange>,
+        thread::JoinHandle<()>,
+        Arc<RwLock<bool>>,
+        Arc<Barrier>,
+    ) {
+        // Create two ends of a boolean variable for signalling when the worker
+        // thread should be terminated
+        let terminate = Arc::new(RwLock::new(false));
+        let terminate_worker = terminate.clone();
+        let barrier = Arc::new(Barrier::new(2));
+        let worker_barrier = barrier.clone();
+
+        // Create a channel for entity changes
+        let (sender, receiver) = channel(100);
+
+        let worker_handle = thread::spawn(move || {
+            // Connect to Postgres
+            let conn = Connection::connect(url, TlsMode::None)
+                .expect("failed to connect entity change listener to Postgres");
+
+            // Obtain a notifications iterator from Postgres
+            let notifications = conn.notifications();
+            let iter = notifications.timeout_iter(Duration::from_millis(500));
+
+            // Subscribe to the "entity_changes" notification channel in Postgres
+            conn.execute("LISTEN entity_changes", &[])
+                .expect("failed to listen to entity changes in Postgres");
+
+            // Wait until the listener has been started
+            barrier.wait();
+
+            let mut notifications = iter.iterator();
+
+            // Read notifications as long as the Postgres connection is alive
+            // or the thread is to be terminated
+            loop {
+                // HACK: Travis seems to have serious problems with running the
+                // integration tests for the Diesel store. Unless we log frequently
+                // from this thread, it all store tests either take a long time
+                // or never finish (before Travis' 10mins timeout).
+                // As soon as you log something from this thread, suddenly all
+                // tests run relatively fast and reliable. This _could_ be a
+                // timing-sensitive synchronization issue on our side we've tried
+                // almost everything and this hack was the only thing that made
+                // the tests work in Travis.
+                if let Ok(_) = env::var("TRAVIS") {
+                    print!(".");
+                    io::stdout().flush().unwrap();
+                }
+
+                // Terminate the thread if desired
+                let terminate_now = terminate.read().unwrap();
+                if *terminate_now {
+                    return;
+                }
+
+                if let Some(Ok(notification)) = notifications.next() {
```

Comment:
This flow would be clearer if where we now have `loop` we had `for notification_result in notifications`. With a `.filter` for the channel name check. Also we should check errors and log them before exiting the loop, right now errors are silent.